afl/kcov_afl.c

```diff
@@ -0,0 +1,236 @@
+
+	/* curlwp can differ for other calls, store the one from open(2) */
+	afl->lid = curlwp->l_lid;
+
+	return afl;
+}
+
+static void
+kcov_afl_free(void *priv)
+{
+	kcov_afl_t *afl;
+
+	afl = (kcov_afl_t *)priv;
+
+	if (afl->afl_area != NULL) {
+		uvm_deallocate(kernel_map, (vaddr_t)afl->afl_area, afl->afl_bsize);
+	}
+
+	kmem_free(afl, sizeof(*afl));
+}
+
+static int
+kcov_afl_setbufsize(void *priv, uint64_t size)
+{
+	kcov_afl_t *afl;
+	int error;
+
+	afl = (kcov_afl_t *)priv;
+
+	if (afl->afl_area != NULL)
+		return EEXIST;
+
+	afl->afl_bsize = size;
+	afl->afl_uobj = uao_create(afl->afl_bsize, 0);
+
+	/* Map the uobj into the kernel address space, as wired. */
+	afl->afl_area = NULL;
+	error = uvm_map(kernel_map, (vaddr_t *)&afl->afl_area, afl->afl_bsize,
+	    afl->afl_uobj, 0, 0, UVM_MAPFLAG(UVM_PROT_RW, UVM_PROT_RW,
+	    UVM_INH_SHARE, UVM_ADV_RANDOM, 0));
+	if (error) {
+		uao_detach(afl->afl_uobj);
+		return error;
+	}
+	error = uvm_map_pageable(kernel_map, (vaddr_t)afl->afl_area,
+	    (vaddr_t)afl->afl_area + afl->afl_bsize, false, 0);
+	if (error) {
+		uvm_deallocate(kernel_map, (vaddr_t)afl->afl_area, afl->afl_bsize);
+		return error;
+	}
+
+	return 0;
+}
+
+static int
+kcov_afl_enable(void *priv, int mode)
+{
+	kcov_afl_t *afl;
+
+	afl = (kcov_afl_t *)priv;
+
+	if (afl->afl_area == NULL)
+		return ENOBUFS;
+
+	return 0;
+}
+
+static int
+kcov_afl_disable(void *priv __unused)
+{
+
+	return 0;
+}
+
+static int
+kcov_afl_mmap(void *priv, size_t size, off_t off, struct uvm_object **uobjp)
+{
+	kcov_afl_t *afl;
+
+	afl = (kcov_afl_t *)priv;
+
+	if ((size + off) > afl->afl_bsize)
+		return ENOMEM;
+
+	uao_reference(afl->afl_uobj);
+
+	*uobjp = afl->afl_uobj;
+
+	return 0;
+}
+
+static void
+kcov_afl_cov_trace_pc(void *priv, intptr_t pc)
+{
+	kcov_afl_t *afl = priv;
+#ifdef HAVE_DEBUGCON_PRINTF
+	debugcon_printf("#: '%x'\n", pc);
+#endif
+	++afl->afl_area[(afl->afl_prev_loc ^ pc) & (afl->afl_bsize-1)];
+	afl->afl_prev_loc = _long_hash64(pc, BITS_PER_LONG);
+
+	return;
+}
+
+static void
+kcov_afl_cov_trace_cmp(void *priv __unused, uint64_t type __unused,
+    uint64_t arg1 __unused, uint64_t arg2 __unused, intptr_t pc __unused)
+{
+
+	/* Not implemented. */
+
+	return;
+}
+
+static struct kcov_ops kcov_afl_ops = {
+	.open = kcov_afl_open,
+	.free = kcov_afl_free,
+	.setbufsize = kcov_afl_setbufsize,
+	.enable = kcov_afl_enable,
+	.disable = kcov_afl_disable,
+	.mmap = kcov_afl_mmap,
+	.cov_trace_pc = kcov_afl_cov_trace_pc,
+	.cov_trace_cmp = kcov_afl_cov_trace_cmp
+};
+
+#ifdef HAVE_DEBUGCON_PRINTF
+MODULE(MODULE_CLASS_MISC, kcov_afl, "kcov,debugcon_printf");
+#else
+MODULE(MODULE_CLASS_MISC, kcov_afl, "kcov");
+#endif
+
+static int
+kcov_afl_modcmd(modcmd_t cmd, void *arg __unused)
+{
+	int error;
+
+	switch (cmd) {
+	case MODULE_CMD_INIT:
+		error = kcov_ops_set(&kcov_afl_ops);
+		if (error)
+			return error;
+		printf("AFL module loaded. ops = %p\n", kcov_afl_ops);
+#ifdef HAVE_DEBUGCON_PRINTF
```

Comment:
In general, I would go for:

```
#ifndef HAVE_DEBUGCON_PRINTF
#define debugcon_printf(a,...)
#endif
```

And use `debugcon_printf` in the code unconditionally.

Reply:
ok good idea